### PR TITLE
Better locale axis time formatting

### PR DESCRIPTION
**Pull request by atmgrifter00**
This change will enable a user to have more control over how the display of time on an axis is formatted, including optionally taking advantage of a browser's language preference for localized display.

### CODE REVIEW

File: jquery.flot.absRelTime.js [hunk 1]
```diff
@@ -19,6 +19,12 @@ xaxis: {
     timeformat: null, // format string to use
     twelveHourClock: false, // 12 or 24 time in time mode
     monthNames: null // list of names of months
+    formatString: "" // A string that allows a user to format the time display (see below for details)
```
[the remark this answers is not in the document]
Comment (atmgrifter00) [reply jquery.flot.absRelTime.js]: I probably should combind formatString into timeFormat. I expect it would then take the form of, as an example:

"%A: #d#T.SS" or "%r: #d#T.SS" (where in the latter case the #d part of the format would be ignored).

The double quotes thing is just me being inconsistent.

File: jquery.flot.absRelTime.js [hunk 2]
```diff
@@ -19,6 +19,12 @@ xaxis: {
...
+    timeEpoch: '0000-12-31T18:00:00' // A time in UTC string format to use as an offset for the display of dates (see below)
+}
+yaxis: {
```
[the remark this answers is not in the document]
Comment (atmgrifter00) [reply jquery.flot.absRelTime.js]: The other options currently exposed on the x-axis are not utilized by my new mechanism, and I saw no compelling reason to worry about synchronizing them right now.

File: jquery.flot.absRelTime.js [hunk 3]
```diff
@@ -38,11 +44,38 @@ The formatted values look like in the example bellow:
...
+There are two flavors of formats you can provide for the date and time formatting options through the
+'formatString' option, and can be mixed between the date and time sections. The first is a simple string
+to indicate date and/or time should be formatted according to the language preference of the browser:
+#d: date
+#T: time
+
+A user could specify formatString = "#d" to indicate the display of a particular date-time should
+just show the date. A string of "#d#T" indicates both date and time should be displayed. Note that time
+is always displayed before the date, and when both are displayed they will be separated by a newline (so order
+of the time and date format substrings doesn't matter).
```
[the remark this answers is not in the document]
Comment (atmgrifter00) [reply jquery.flot.absRelTime.js]: I would call it a temporary limitation, but is consistent with current behavior. It might not be that hard to address now (I'll take a look).

File: jquery.flot.absRelTime.js [hunk 4]
```diff
@@ -38,11 +44,38 @@ The formatted values look like in the example bellow:
...
+hh(or HH):mm   (HH will indicate that the formatted time should be in 24-hour format)
+hh:mm:ss
+hh:mm:ss(.S+) (where 'S+' just means you can specify a particular number of 'S's for fractional seconds)
+
 ### Relative time axis
 A relative time axis will show the time values with respect to the first data sample.
 Basically, the first datapoint from the points array will be considered time 00:00:00:00.
 If the difference between two datapoints is small, the milliseconds will apear.
 Otherwise, the time representation will contain only the hour, minute and second.
+NOTE: If a formatString is provided, any date format will be ignored.
```
[the remark this answers is not in the document]
Comment (atmgrifter00) [reply jquery.flot.absRelTime.js]: Perhaps we don't, but I don't think I need to address that now.

File: jquery.flot.absRelTime.js [hunk 5]
```diff
@@ -153,15 +331,33 @@ the the second one the date in gregorian date format.
...
-            result += leftPad(minutes) + ':';
-            result += leftPad(seconds);
-            if (showMilliseconds) {
-                result += '.' + leftPadNTimes(milliseconds, "0", 3);
+            result += leftPad(minutes);
+            if (showSeconds) {
+                result += (":" + leftPad(seconds));
+            }
+
+            var forceMilliseconds = showSeconds && formatString.includes(".S");
```
[the remark this answers is not in the document]
Comment (atmgrifter00) [reply jquery.flot.absRelTime.js]: I'll look into it.

File: jquery.flot.absRelTime.js [hunk 6]
```diff
@@ -153,15 +331,33 @@ the the second one the date in gregorian date format.
...
+            showMilliseconds = (showMilliseconds && formatString === "") || forceMilliseconds;
+            var padAmount = 3;
+            if (forceMilliseconds) {
+                var fractionalSecondsIndex = formatString.indexOf(".S");
+                var numberOfFractionalSeconds = fractionalSecondsIndex > 0 ? formatString.substring(fractionalSecondsIndex + 1) : 0;
```
[the remark this answers is not in the document]
Comment (atmgrifter00) [reply jquery.flot.absRelTime.js]: Whoops. You found the spot where I didn't handle the case where that format string may not be the last part of the string. Thus, the need for a helper mechanism.

File: tests/jquery.flot-absoluteTimeAxis.Test.js [hunk 7]
```diff
@@ -145,6 +149,127 @@ describe('A Flot chart with absolute time axes', function () {
         ]);
     });
 
+    it('shows only hours and minutes with formatString: "hh:mm"', function () {
```
[the remark this answers is not in the document]
Comment (atmgrifter00) [reply tests/jquery.flot-absoluteTimeAxis.Test.js]: No, and I should.

File: jquery.flot.absRelTime.js [hunk 8]
```diff
@@ -127,15 +299,21 @@ the the second one the date in gregorian date format.
...
             }
 
-            var gregorianDate = makeUtcWrapper(new Date(date - unixToAbsoluteEpochDiff)).date;
+            var gregorianDate = makeUtcWrapper(new Date(date.valueOf() + timeEpoch)).date;
+
+            var time;
+            if (formatString !== "") {
+                time = getFormattedDateTimeString(gregorianDate, formatString);
+            } else {
+                var msString = showMilliseconds ? '.' + padNTimes(ms, '0', 3) : '';
```
[the remark this answers is not in the document]
Comment (atmgrifter00) [reply jquery.flot.absRelTime.js]: Unfortunately no. The new mechanism being used to format the date (Intl.DateTimeFormat) will produce a different representation for year 0000, then the current mechanism. Preserving the else case is strictly to support the old behavior, despite it being pretty narrow, and possibly useless.

[the remark this answers is not in the document]
Comment (EmperorXLII) [reply jquery.flot.absRelTime.js]: Do you mean "year 0001"?  (There is no "year zero" — and if we are trying to pretend to support that, I think we can dispense with the else :)

File: jquery.flot.absRelTime.js [hunk 9]
```diff
@@ -44,32 +62,6 @@ The formatted values look like in the example bellow:
...
-just show the date. A string of "#d#T" indicates both date and time should be displayed. Note that time
-is always displayed before the date, and when both are displayed they will be separated by a newline (so order
-of the time and date format substrings doesn't matter).
-
-The second flavor is a more verbose set of options, allowing a user to specify certain characteristics of both
-the date and time displays. The following shows the complete set of options:
-dd/MM/yy(yy) hh:mm:ss.SSS  (NOTE: The (yy) just indicates you can optionally specify a 4 digit year. So 'yy' or 'yyyy'.)
-
-Acceptable date configutations are:
-dd/MM</yy(yy)>  <...> indicates optional part of formatString
```
Comment (EmperorXLII) [inline jquery.flot.absRelTime.js]: I was using `<...>` to indicate optional values, as you did here (to avoid ambiguity with choices between required alternatives `(...|...)`).  Are you trying to say every format above is optional?

File: tests/jquery.flot-absoluteTimeAxis.Test.js [hunk 10]
```diff
@@ -145,6 +148,127 @@ describe('A Flot chart with absolute time axes', function () {
...
+        expect(firstAndLast(plot.getAxes().xaxis.ticks)).toEqual([
+            {v: 0 + tzDiff, label: '12:00:00.00 AM'},
+            {v: 60 + tzDiff, label: '12:01:00.00 AM'}
+        ]);
+    });
+
+    // TODO: Figure out the proper way to test localized strings of languages other than "en-us".
+    // You can't set navigator.language to subvert the desired browser language. This is what the
+    // implementation is using to localize.
+    xit('shows localized date time properly with "#T#d" formatString"', function () {
```
[the remark this answers is not in the document]
Comment (atmgrifter00) [reply tests/jquery.flot-absoluteTimeAxis.Test.js]: Yes, if only to demonstrate the type of test I think we want to have.

File: tests/jquery.flot-relativeTimeAxis.Test.js [hunk 11]
```diff
@@ -186,4 +187,31 @@ describe('A Flot chart with relative time axes', function () {
...
+            {v: 0, label: '00:00:00'},
+            {v: 61, label: '00:01:01'}
+        ]);
+    });
+
+    it('shows proper number of millisecond digits with formatString: "hh:mm:ss.SSS"', function () {
+        plot = createPlotWithRelativeTimeAxis(placeholder, [[[0 + 0.5, 1], [61 + 0.001, 2]]], "hh:mm:ss.SSS");
+
+        expect(firstAndLast(plot.getAxes().xaxis.ticks)).toEqual([
+            {v: 0 + 0.5, label: '00:00:00.500'},
```
[the remark this answers is not in the document]
Comment (atmgrifter00) [reply tests/jquery.flot-relativeTimeAxis.Test.js]: I would have thought so, but note that the existing tests have the same behavior. Also, I changed nothing about the mechanism being used to format the time in the relative case, just which parts to show. This might merit later investigation.

File: jquery.flot.absRelTime.js [hunk 12]
```diff
@@ -113,7 +150,139 @@ the the second one the date in gregorian date format.
...
+        }
+
+        function getFormattedTimeString(date, formatString, formatOptions, locale) {
+            var showTime = formatString.includes("hh") || formatString.includes("HH") || formatString.includes("#T");
+            if (!showTime) {
+                return "";
+            }
+
+            var fractionalSecondsString = getFractionalSecondsString(date.getMilliseconds(), formatString);
+            var numberOfFractionalSeconds = fractionalSecondsString.length;
```
Comment (EmperorXLII) [inline jquery.flot.absRelTime.js]: This fractional seconds count variable is declared very far away from its actual use, and it doesn't add much over accessing the string length directly (where it is used to determine if we need to append the string).

File: jquery.flot.absRelTime.js [hunk 13]
```diff
@@ -182,9 +360,16 @@ the the second one the date in gregorian date format.
             var c = fmt.charAt(i),
                 localDateValue = d.date || d.getDate();
             if (escape) {
+                var timeFormatStartIndex = fmt.indexOf('<');
```
Comment (EmperorXLII) [inline jquery.flot.absRelTime.js]: Oh, so the angle brackets _are_ a required part of the format string!  Please disregard my earlier doc comment.
